nixos/searx: use envsubst instead of sed for reading environment variables

codebam opened the pull request:
## Things done

- use envsubst instead of sed for reading environment variables
- update usage examples
- add release notes
- tested with agenix

The current implementation using sed fails to parse the password "GrdmPy~&?:VV=k.#B2R8er(]7"

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

acid-bong commented:
Since you edited the module, the PR and the commit should be prefixed with `nixos/searx`

acid-bong commented on nixos/modules/services/networking/searx.nix:
```diff
@@ -90,12 +89,12 @@ in
           { server.port = 8080;
             server.bind_address = "0.0.0.0";
-            server.secret_key = "@SEARX_SECRET_KEY@";
+            server.secret_key = "$\{SECRET_KEY}";
```
```suggestion
            server.secret_key = "''${SECRET_KEY}";
```
In multiline strings, ${} is escaped with two apostrophes. Same below

codebam replied:
Okay thank you

SuperSandro2000 commented on nixos/modules/services/networking/searx.nix:
```diff
@@ -90,12 +89,12 @@ in
               { name = "wolframalpha";
                 shortcut = "wa";
-                api_key = "@WOLFRAM_API_KEY@";
+                api_key = "''${API_KEY}";
```
```suggestion
                api_key = "''${WOLFRAM_API_KEY}";
```

SuperSandro2000 commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -49,6 +49,9 @@
 - `reaction` has been updated to version 2, which includes some breaking changes.
   For more information, [check the release article](https://blog.ppom.me/en-reaction-v2).
 
+- `searx` was updated to use `envsubst` instead of `sed` for parsing config environment variables.
```
What do we need to migrate? To me it looks like things stay compatible.

rnhmjoj replied:
It a breaking change: `@SECRET@` → `$SECRET`.

rnhmjoj replied:
This could be explained directly here, though. With a couple of examples.

SuperSandro2000 replied:
That definitely needs to be explained like this. I didn't even grasp it when first time reading it.

codebam replied:
I added one example of @SECRET@ to $SECRET. If you think we need more I can add another.

rnhmjoj commented on nixos/modules/services/networking/searx.nix:
```diff
@@ -90,12 +89,12 @@ in
           { server.port = 8080;
             server.bind_address = "0.0.0.0";
-            server.secret_key = "@SEARX_SECRET_KEY@";
+            server.secret_key = "''${SECRET_KEY}";
```
envsubs seems to support `$variable` as well. I would use that to avoid the escaping mess with Nix antiquotation.

codebam replied:
How does envsubs support `$variable`? I think this would be better than `${variable}` as well

rnhmjoj replied:
> How does envsubs support $variable?

I'm not sure what you're asking. It seems to support both `${variable}` and `$variable` syntaxes, see https://github.com/a8m/envsubst?tab=readme-ov-file#docs.

codebam replied:
Oh I see. Thanks!

codebam committed:
nixos/searx: use envsubst instead of sed for reading environment variables

- use envsubst instead of sed for reading in environment variables
- update usage examples
- add release notes

SuperSandro2000 commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -66,6 +66,9 @@
   The modern alternative is to configure `pyproject = true` with `build-system = [ setuptools ]`.
 
+- `searx` was updated to use `envsubst` instead of `sed` for parsing config environment variables.
+  For information on how to migrate see the [usage](https://github.com/a8m/envsubst) of `envsubst` and view the updated examples.
```
```suggestion
```
I don't think that this is a detail people need to know but I like the example idea from rnhmjoj very much and would rather want to have that here instead.

We should probably use the `server.secret_key = "@SEARX_SECRET_KEY@"` setting as an example as I expect most if not all configurations to use that.

codebam replied:
Did this, please let me know if you have any more suggestions on this. Thanks!

SuperSandro2000 commented:
Also please only merge this after #346777 got merged. I do not really want to rebase that again 😅

rnhmjoj reviewed (approved):
It would be useful to check for the old syntax (maybe using `grep -q` in the `preStart` script) and error out. But only if you have time.

Otherwise, the changes looks good and the tests are ok.

rnhmjoj commented:
All right, let's merge this.
